### ISSUE TITLE
diag_indices function issue?

### ISSUE THREAD

**Issue by myakhlakov** [numpy/numpy#15681]
https://github.com/numpy/numpy/blob/d9b1e32cb8ef90d6b4a47853241db2a28146a57d/numpy/lib/index_tricks.py#L883-L949

Hello there! 
Is it ok that `diag_indices` function returns two links on one index array?

For example, I'm trying to use this function to acces not only to the main diagonal but also to the diagonal with the given offset (like using `offset` parameter in `np.diagonal`). To implement that I do something like this:
```
X = np.array(...) # some 2D array
n, m = X.shape # without loss of generality assume that n <= m
xi, yi = np.diag_indices(n)
yi += offset
xi, yi = xi[np.logical_and(yi < m,yi >= 0)], yi[np.logical_and(yi < m, yi >= 0)]

print(xi, yi) # print to check if values are different (they aren't)
print(xi is yi) # print to check out if instances are different (they aren't)
```

To fix this issue I suggest to do the following:

```
...
return list(arange(n) for i in range(ndim))
```

**Comment by rossbar** [comment]
If I'm understanding this correctly, the issue you're having is related to the fact that the index arrays (`xi` and `yi` in your example) refer to the same values in memory, correct? In other words, you'd like to be able to modify `yi` without affecting `xi`.

If that's the case, rather than modifying `np.diag_indices` itself, why not create a copy:
```python
>>> xi, yi = np.diag_indices(n)
>>> yi = yi.copy()
>>> xi is yi
False
```

**Comment by myakhlakov** [comment]
> If I'm understanding this correctly, the issue you're having is related to the fact that the index arrays (`xi` and `yi` in your example) refer to the same values in memory, correct? In other words, you'd like to be able to modify `yi` without affecting `xi`.
>
> If that's the case, rather than modifying `np.diag_indices` itself, why not create a copy:

Yes, formally you're correct.  
But I want to make clear some points why refering to the same part of memory isn't great.

1.  it's unclear from the function description that index arrays are refering to the same values in memory;
2. it's unconvenient. formally `xi` and `yi` are not the same: `xi` refers to indices at the first dimension,  `yi` -- at the second;
3. there is no function in numpy which could give us not "read-only" access to the diagonal of the array (with offset too) -- `np.diagonal` returns the copy of the corresponding diagonal instead. Than anyone (not only me) could try to find the function which returns indices of the diagonal instead and `np.diag_indices` is obvious solution in that case. Reffering again to the point 1, it's quite surprising that during change of indices over one dimension you also change them over every over dimension too.  After all, this functions is not so memory-hungry to make such rude optimization.


P.S. Sorry for my english in advance. Feel free to ask again if necessary

**Comment by seberg** [comment]
Hmm, this is a bit strange. I agree with you @myakhlakov, although the function also does not have any `offset` arguments similar to `np.diag`. If there was an `offset` argument, then it would be more obvious that there should not be a difference in behaviour depending on whether `offset` is given or not.

**Comment by rossbar** [comment]
After a closer look, `diag_indices` falls into the category of a "convenience" function that, in my opinion, is actually *inconvenient*. The operation is a two-liner:

https://github.com/numpy/numpy/blob/c31cc36a8a814ed4844a2a553454185601914a5a/numpy/lib/index_tricks.py#L946-L947

I'm not against modifying it to return independent copies as it's hard to think of a use-case where the memory-saving is critical.

OTOH - I would be even more positive about getting the ball rolling on removing it entirely. The benefit of having `[np.arange(n)] * ndim` wrapped up in it's own function call hardly seems useful (and indeed, as this issue attests, can actually lead to *more* problems due to the obfuscation).

@seberg in your opinion, is this something we could take to the mailing list? A quick search shows no hits for matplotlib or pandas, and only two uses in scipy.

**Comment by eric-wieser** [comment]
> The benefit of having `[np.arange(n)] * ndim` wrapped up in it's own function call

Note that `(np.arange(n),) * ndim` is different in an important way from what you say there.

Agreed that this function is inconvenient though.

One thing we could do here is make the return value warn-on-write. Another would be to make the return value read-only.

I wonder if we can finally make `np.diagonal` return a read-write view, so that the need for this function vanishes.

**Comment by seberg** [comment]
Yeah, the diagonal thing has been warning since forever (at least it feels like it). I am happy with considering a full deprecation, or going the write-only/warn-on-write way.
Going to the mailing list (best with a preferred/suggested choice maybe) sounds like a good idea.
